api/test/unit/management/commands/generatewaveforms_test.py

```diff
@@ -0,0 +1,177 @@
+import subprocess
+from io import StringIO
+from test.factory.faker import WaveformProvider
+from test.factory.models.audio import AudioAddOnFactory, AudioFactory
+from unittest import mock
+
+import psycopg2
+import pytest
+from catalog.api.models.audio import Audio, AudioAddOn
+from django.core.management import call_command
+from django.db import connections
+from django.test.utils import CaptureQueriesContext
+
+
+@mock.patch("catalog.api.models.audio.generate_peaks")
+def call_generatewaveforms(mock_generate_peaks: mock.MagicMock) -> tuple[str, str]:
+    mock_generate_peaks.side_effect = lambda _: WaveformProvider.generate_waveform()
+    out = StringIO()
+    err = StringIO()
+    call_command("generatewaveforms", unlimited=True, stdout=out, stderr=err)
+
+    return out.getvalue(), err.getvalue()
+
+
+def assert_all_audio_have_waveforms():
+    assert (
+        list(
+            AudioAddOn.objects.filter(waveform_peaks__isnull=False).values_list(
+                "audio_identifier"
+            )
+        ).sort()
+        == list(Audio.objects.all().values_list("identifier")).sort()
+    )
+
+
+@pytest.mark.django_db
+def test_creates_waveforms_for_audio():
+    AudioFactory.create_batch(153)
+
+    assert AudioAddOn.objects.count() == 0
+
+    call_generatewaveforms()
+
+    assert_all_audio_have_waveforms()
+
+
+@pytest.mark.django_db
+def test_does_not_reprocess_existing_waveforms():
+    waveformless_audio = AudioFactory.create_batch(3)
+
+    # AudioAddOnFactory will create associated Audio objects as well
+    # so those three will serve as the audio that should _not_ get processed
+    AudioAddOnFactory.create_batch(3)
+
+    # Create an add on that doesn't have a waveform, this one should get processed as well
+    null_waveform_addon = AudioAddOnFactory.create(waveform_peaks=None)
+    waveformless_audio.append(
+        Audio.objects.get(identifier=null_waveform_addon.audio_identifier)
+    )
+
+    out, err = call_generatewaveforms()
+
+    assert f"Generating waveforms for {len(waveformless_audio)} records" in out
+    assert_all_audio_have_waveforms()
+
+
+@pytest.mark.django_db
+@mock.patch("catalog.api.models.audio.generate_peaks")
+def test_paginates_audio_waveforms_to_generate(
+    mock_generate_peaks, django_assert_num_queries
+):
+    mock_generate_peaks.return_value = WaveformProvider.generate_waveform()
+
+    audio_count = 53  # 6 pages
+    pages = 6
+    AudioFactory.create_batch(audio_count)
+
+    test_audio = AudioFactory.create()
+    with CaptureQueriesContext(connections["default"]) as capture:
+        test_audio.get_or_create_waveform()
+    test_audio.delete()
+
+    queries_per_iteration = len(capture.captured_queries)
+
+    pagination_queries = pages + 1  # 1 per page + the final empty page's query
+    count_queries = 1  # initializes the count for tqdm
+    interation_queries = (
+        queries_per_iteration * audio_count
+    )  # queries inside get_or_create_waveform
```

Comment:
I know the format isn't great as a result of this change, but if these comments are on the line above, `black` won't try and split L87-L89 over multiple lines

Reply:
Done in 8fb6223843372f6714aaea31dc1ea0545751c9f7